File: math/fastex/fastex.c


long long recursiveFastex(int b, long long e)
{
    if(!e) return 1;
    long long r = fastex(b,e >>1);
    return (e &1 ? r*r*b : r*r) %1000000007;
}


long long iterativeFastex(long long b, long long e)
{
    long long r = 1;

    for(; e; e >>= 1)
    {
        if(e &1) r = b*r;
        b = b*b;
    }
    return r;
}

File: math/fastex/fastex.py


def recursive_fastex(b :int, e :int) -> int :

    if not e : return 1
    r = recursive_fastex(b,e >>1)
    return (b*r*r if e &1 else r*r)


def iterative_fastex(b :int, e :int) -> int :

    r = 1

    while   0 <e:
        if  e &1 : r = (r*b)

        b = (b*b)
        e >>= 1

    return r

